server/src/main/java/de/uftos/repositories/ucdl/parser/DefinitionParser.java

```diff
@@ -0,0 +1,691 @@
+      }
+      case "ELEMENTNAME" -> {
+        String image = ((SimpleNode) root).jjtGetFirstToken().image;
+        if (!parameters.containsKey(image)) {
+          throw new ParseException("Parameter/Variable \"" + image + "\" does not exist!");
+        }
+        return new ValueDto<>(UcdlToken.VALUE_REFERENCE, image);
+      }
+      case "NUMBERELEMENT" -> {
+        if (((SimpleNode) root).jjtGetFirstToken().image.equals("size")) {
+          List<AbstractSyntaxTreeDto> set = new ArrayList<>();
+          set.add(buildAst(root.jjtGetChild(0), parameters));
+          return new OperatorDto(UcdlToken.SIZE, set);
+        } else {
+          return new ValueDto<>(UcdlToken.NUMBER, Integer.parseInt(
+              ((SimpleNode) root).jjtGetFirstToken().image));
+        }
+      }
+      case "SET" -> {
+        AbstractSyntaxTreeDto setName = buildAst(root.jjtGetChild(0).jjtGetChild(0), parameters);
+
+        ResourceType setType;
+
+        if (setName.getToken() == UcdlToken.NUMBER_SET) {
+          setType = ResourceType.NUMBER;
+        } else {
+          setType = ((ElementDto) setName).type();
+        }
+
+        List<AbstractSyntaxTreeDto> modifiers = new ArrayList<>();
+
+        Node modifierList = root.jjtGetChild(1);
+        while (modifierList.jjtGetNumChildren() > 0) {
+          if (modifierList.jjtGetNumChildren() == 2) {
+            SimpleNode flatmap = (SimpleNode) modifierList.jjtGetChild(0);
+            String attribute = flatmap.jjtGetFirstToken().image;
+            setType = getNextResourceType(setType, attribute);
+            modifiers.add(new ValueDto<>(UcdlToken.ATTRIBUTE, attribute));
+            modifierList = modifierList.jjtGetChild(1);
+
+          } else if (modifierList.jjtGetNumChildren() == 3) {
+            ResourceType thisType = parameters.get("this");
+            parameters.replace("this", setType);
+
+            List<AbstractSyntaxTreeDto> filters = new ArrayList<>();
+            filters.add(buildAst(modifierList.jjtGetChild(0), parameters));
+
+            Node filterList = modifierList.jjtGetChild(1);
+            while (filterList.jjtGetNumChildren() > 0) {
+              filters.add(buildAst(filterList.jjtGetChild(0), parameters));
+              filterList = filterList.jjtGetChild(1);
+            }
+
+            for (AbstractSyntaxTreeDto filter : filters) {
+              if ((filter.getToken() == UcdlToken.NUMBER_SET && setType != ResourceType.NUMBER)
+                  || (filter.getToken() == UcdlToken.RESOURCE_SET
+                  && ((SetDto) filter).type() != setType)) {
+                throw new ParseException("Sets can only be filtered using sets of the same type!");
+              }
+            }
+
+            modifiers.add(new OperatorDto(UcdlToken.FILTER, filters));
+            modifierList = modifierList.jjtGetChild(2);
+            parameters.replace("this", thisType);
+          }
+        }
```

Comment:
I think this should get a massive cleanup. What about cases where `#children` is one or greater than 3? Can that never happen?

Reply:
The amount of children is directly tied to the grammar.
So no, if the AST given was parsed using the javacc-parser than it is not possible to have different a differen amount of children than 0, 2 or 3. 
(The specification in the grammar file: 
void SETMODIFICATION(): {} { <FLATMAP> ATTRIBUTE() SETMODIFICATION() | <LBRACKET> FILTER() FILTERLIST() <RBRACKET> SETMODIFICATION() | {} }
)